File: android/app/src/main/java/com/jackbarkerapps/openkeep/security/KeyManager.kt
package com.jackbarkerapps.openkeep.security

import android.content.Context
import android.content.SharedPreferences
import android.util.Base64
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import java.security.SecureRandom
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.PBEKeySpec
import javax.crypto.Cipher
import javax.crypto.spec.SecretKeySpec
import javax.crypto.spec.GCMParameterSpec

class KeyManager(private val context: Context) {

    companion object {
        private const val SHARED_PREFS_FILENAME = "secure_prefs"
        private const val KEY_ALIAS = "db_master_key"
        private const val ENCRYPTED_MASTER_KEY_V2 = "encrypted_master_key_v2"
        private const val SALT_KEY = "kdf_salt"
        private const val SALT_SIZE = 16
        private const val ITERATIONS = 10000 // OWASP recommends higher, but this is reasonable for mobile
        private const val KEY_LENGTH = 256
    }

    private val masterKey by lazy {
        MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
    }

    private val securePrefs: SharedPreferences by lazy {
        try {
            createSecurePrefs()
        } catch (e: Exception) {
            android.util.Log.e("KeyManager", "Failed to create EncryptedSharedPreferences, retrying after delete", e)
            try {
                // Delete the file and try again - this handles cases where the key is lost/corrupted
                deletePrefs()
                createSecurePrefs()
            } catch (e2: Exception) {
                android.util.Log.e("KeyManager", "Secondary failure creating EncryptedSharedPreferences, attempting Keystore clear", e2)
                try {
                    // Last ditch effort: clear the Keystore entry itself
                    clearKeyStore()
                    deletePrefs()
                    createSecurePrefs()
                } catch (e3: Exception) {
                    android.util.Log.e("KeyManager", "Fatal error creating EncryptedSharedPreferences after Keystore clear", e3)
                    throw e3
                }
            }
        }
    }

    private fun deletePrefs() {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
            context.deleteSharedPreferences(SHARED_PREFS_FILENAME)
        } else {
            val file = java.io.File(context.filesDir.parent, "shared_prefs/$SHARED_PREFS_FILENAME.xml")
            if (file.exists()) {
                file.delete()
            }
        }
    }

    private fun clearKeyStore() {
        try {
            val keyStore = java.security.KeyStore.getInstance("AndroidKeyStore")
            keyStore.load(null)
            keyStore.deleteEntry(MasterKey.DEFAULT_MASTER_KEY_ALIAS)
            android.util.Log.d("KeyManager", "Cleared MasterKey from KeyStore")
        } catch (e: Exception) {
            android.util.Log.e("KeyManager", "Failed to clear KeyStore entry", e)
        }
    }

    private fun createSecurePrefs(): SharedPreferences {
        return EncryptedSharedPreferences.create(
            context,
            SHARED_PREFS_FILENAME,
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        )
    }

    private val standardPrefs: SharedPreferences by lazy { 
        context.getSharedPreferences("app_prefs", Context.MODE_PRIVATE)
    }

    fun getOrGenerateSalt(): ByteArray {
        val saltString = standardPrefs.getString(SALT_KEY, null)
        if (saltString != null) {
            return Base64.decode(saltString, Base64.DEFAULT)
        }

        val salt = ByteArray(SALT_SIZE)
        SecureRandom().nextBytes(salt)
        val encodedSalt = Base64.encodeToString(salt, Base64.DEFAULT)
        standardPrefs.edit().putString(SALT_KEY, encodedSalt).apply()
        return salt
    }

    // Derives local KEK using local salt
    fun deriveLocalKEK(pin: String): ByteArray {
        val salt = getOrGenerateSalt()
        val spec = PBEKeySpec(pin.toCharArray(), salt, ITERATIONS, KEY_LENGTH)
        val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
        return factory.generateSecret(spec).encoded
    }

    // Derives Export KEK using static salt for cloud syncing
    private fun deriveExportKEK(pin: String): ByteArray {
        // A static 16-byte salt solely for wrapping the Master Key for cloud export
        val staticSalt = "OpenKeepCloudExportSalt123".toByteArray(Charsets.UTF_8).copyOf(16)
        val spec = PBEKeySpec(pin.toCharArray(), staticSalt, ITERATIONS, KEY_LENGTH)
        val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
        return factory.generateSecret(spec).encoded
    }

    private fun encryptWithKEK(payload: ByteArray, kek: ByteArray): String {
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        val spec = SecretKeySpec(kek, "AES")
        cipher.init(Cipher.ENCRYPT_MODE, spec)
        val iv = cipher.iv
        val ciphertext = cipher.doFinal(payload)
        val combined = ByteArray(iv.size + ciphertext.size)
        System.arraycopy(iv, 0, combined, 0, iv.size)
        System.arraycopy(ciphertext, 0, combined, iv.size, ciphertext.size)
        return Base64.encodeToString(combined, Base64.DEFAULT)
    }

    private fun decryptWithKEK(encryptedBase64: String, kek: ByteArray): ByteArray {
        val combined = Base64.decode(encryptedBase64, Base64.DEFAULT)
        val iv = ByteArray(12)
        if (combined.size < 12) throw IllegalArgumentException("Invalid encrypted payload")
        System.arraycopy(combined, 0, iv, 0, 12)
        val ciphertext = ByteArray(combined.size - 12)
        System.arraycopy(combined, 12, ciphertext, 0, ciphertext.size)
        
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        val spec = SecretKeySpec(kek, "AES")
        val gcmSpec = GCMParameterSpec(128, iv)
        cipher.init(Cipher.DECRYPT_MODE, spec, gcmSpec)
        return cipher.doFinal(ciphertext)
    }

    fun hasV2Key(): Boolean {
        return securePrefs.contains(ENCRYPTED_MASTER_KEY_V2)
    }

    fun generateRandomMasterKeyAndSave(pin: String): ByteArray {
        val masterKey = ByteArray(32)
        SecureRandom().nextBytes(masterKey)
        val kek = deriveLocalKEK(pin)
        val encrypted = encryptWithKEK(masterKey, kek)
        securePrefs.edit().putString(ENCRYPTED_MASTER_KEY_V2, encrypted).apply()
        return masterKey
    }

    fun getMasterKeyForPin(pin: String): ByteArray {
        val v2Encoded = securePrefs.getString(ENCRYPTED_MASTER_KEY_V2, null)
        if (v2Encoded != null) {
            val kek = deriveLocalKEK(pin)
            return decryptWithKEK(v2Encoded, kek)
        }
        // Fallback to V1 logic where the derived PIN key IS the Master Key
        return deriveLocalKEK(pin)
    }

    fun upgradeToV2(masterKey: ByteArray, pin: String) {
        val v2Encoded = securePrefs.getString(ENCRYPTED_MASTER_KEY_V2, null)
        if (v2Encoded == null) {
            val kek = deriveLocalKEK(pin)
            val encrypted = encryptWithKEK(masterKey, kek)
            securePrefs.edit().putString(ENCRYPTED_MASTER_KEY_V2, encrypted).apply()
        }
    }

    fun exportCloudMasterKey(pin: String): String {
        val masterKey = getMasterKeyForPin(pin)
        val exportKEK = deriveExportKEK(pin)
        return encryptWithKEK(masterKey, exportKEK)
    }

    fun importCloudMasterKey(payload: String, pin: String) {
        val exportKEK = deriveExportKEK(pin)
        val masterKey = decryptWithKEK(payload, exportKEK)
        // Save it locally wrapped with local KEK
        val localKEK = deriveLocalKEK(pin)
        val encryptedLocal = encryptWithKEK(masterKey, localKEK)
        securePrefs.edit().putString(ENCRYPTED_MASTER_KEY_V2, encryptedLocal).apply()
        // Also save it for auto-unlock
        storeMasterKey(masterKey)
    }

    fun canDecryptCloudMasterKey(payload: String, pin: String): Boolean {
        return try {
            val exportKEK = deriveExportKEK(pin)
            decryptWithKEK(payload, exportKEK)
            true
        } catch (e: Exception) {
            false
        }
    }

    fun verifyCloudMasterKeyMatch(payload: String, pin: String): Boolean {
        try {
            val exportKEK = deriveExportKEK(pin)
            val cloudMasterKey = decryptWithKEK(payload, exportKEK)
            val currentMasterKey = getMasterKeyForPin(pin)
            return cloudMasterKey.contentEquals(currentMasterKey)
        } catch (e: Exception) {
            return false
        }
    }

    fun changePinV2(oldPin: String, newPin: String) {
        val masterKey = getMasterKeyForPin(oldPin)
        val newLocalKEK = deriveLocalKEK(newPin)
        val encryptedLocal = encryptWithKEK(masterKey, newLocalKEK)
        securePrefs.edit().putString(ENCRYPTED_MASTER_KEY_V2, encryptedLocal).apply()
    }

    fun storeMasterKey(key: ByteArray) {
        val encodedKey = Base64.encodeToString(key, Base64.DEFAULT)
        securePrefs.edit().putString(KEY_ALIAS, encodedKey).apply()
    }

    fun getMasterKey(): ByteArray? {
        val encodedKey = securePrefs.getString(KEY_ALIAS, null) ?: return null
        return Base64.decode(encodedKey, Base64.DEFAULT)
    }

    fun encrypt(plaintext: String, key: ByteArray? = getMasterKey()): String {
            if (key == null) throw IllegalStateException("No master key available")
            
            val cipher = javax.crypto.Cipher.getInstance("AES/GCM/NoPadding")
            val spec = javax.crypto.spec.SecretKeySpec(key, "AES")
            cipher.init(javax.crypto.Cipher.ENCRYPT_MODE, spec)
            
            val iv = cipher.iv
            val ciphertext = cipher.doFinal(plaintext.toByteArray(Charsets.UTF_8))
            
            // Return IV + Ciphertext encoded in Base64
            val combined = ByteArray(iv.size + ciphertext.size)
            System.arraycopy(iv, 0, combined, 0, iv.size)
            System.arraycopy(ciphertext, 0, combined, iv.size, ciphertext.size)
            
            // NO_WRAP prevents line breaks in base64 output, which avoids
            // issues when the encrypted data is transmitted via JSON (e.g. Google Drive sync).
            return Base64.encodeToString(combined, Base64.NO_WRAP)
        }

    fun decrypt(encryptedData: String, key: ByteArray? = getMasterKey()): String {
        if (key == null) throw IllegalStateException("No master key available")

        val combined = Base64.decode(encryptedData, Base64.DEFAULT)
        
        // GCM IV is usually 12 bytes
        val iv = ByteArray(12)
        if (combined.size < 12) throw IllegalArgumentException("Invalid encrypted data")
        System.arraycopy(combined, 0, iv, 0, 12)
        
        val ciphertext = ByteArray(combined.size - 12)
        System.arraycopy(combined, 12, ciphertext, 0, ciphertext.size)
        
        val cipher = javax.crypto.Cipher.getInstance("AES/GCM/NoPadding")
        val keySpec = javax.crypto.spec.SecretKeySpec(key, "AES")
        val gcmSpec = javax.crypto.spec.GCMParameterSpec(128, iv) // 128 bit auth tag
        
        cipher.init(javax.crypto.Cipher.DECRYPT_MODE, keySpec, gcmSpec)
        val plaintextBytes = cipher.doFinal(ciphertext)
        
        return String(plaintextBytes, Charsets.UTF_8)
    }

    fun clear() {
        securePrefs.edit().remove(KEY_ALIAS).apply()
        // We do NOT clear ENCRYPTED_MASTER_KEY_V2, as that represents our locked database key.
        // Clearing KEY_ALIAS simply "locks" the auto-unlock feature.
    }

    fun clearAll() {
        clear()
        securePrefs.edit().remove(ENCRYPTED_MASTER_KEY_V2).apply()
    }
}

File: ios/App/App/KeyManager.swift
import Foundation
import CryptoKit
import CommonCrypto
import Security

class KeyManager {
    
    private let KEY_ALIAS = "db_master_key"
    private let ENCRYPTED_MASTER_KEY_V2 = "encrypted_master_key_v2"
    private let SALT_KEY = "kdf_salt"
    private let SALT_SIZE = 16
    private let ITERATIONS: UInt32 = 10000
    private let KEY_LENGTH = 32 // 256 bits
    
    // UserDefaults used for salt, like Android's standardPrefs
    private let defaults = UserDefaults.standard
    
    init() {}
    
    func getOrGenerateSalt() -> [UInt8] {
        if let saltString = defaults.string(forKey: SALT_KEY),
           let saltData = Data(base64Encoded: saltString) {
            return [UInt8](saltData)
        }
        
        var salt = [UInt8](repeating: 0, count: SALT_SIZE)
        _ = SecRandomCopyBytes(kSecRandomDefault, SALT_SIZE, &salt)
        
        let encodedSalt = Data(salt).base64EncodedString()
        defaults.set(encodedSalt, forKey: SALT_KEY)
        return salt
    }
    
    func deriveLocalKEK(pin: String) throws -> [UInt8] {
        let salt = getOrGenerateSalt()
        return try pbkdf2(password: pin, salt: salt, iterations: ITERATIONS, keyLength: KEY_LENGTH)
    }
    
    func deriveExportKEK(pin: String) throws -> [UInt8] {
        // "OpenKeepCloudExportSalt123" first 16 bytes
        let staticSaltStr = String("OpenKeepCloudExportSalt123".prefix(16))
        let staticSalt = [UInt8](staticSaltStr.data(using: .utf8)!)
        return try pbkdf2(password: pin, salt: staticSalt, iterations: ITERATIONS, keyLength: KEY_LENGTH)
    }
    
    private func pbkdf2(password: String, salt: [UInt8], iterations: UInt32, keyLength: Int) throws -> [UInt8] {
        guard let passwordData = password.data(using: .utf8) else {
            throw NSError(domain: "KeyManager", code: 1, userInfo: [NSLocalizedDescriptionKey: "Invalid PIN"])
        }
        
        var derivedKey = [UInt8](repeating: 0, count: keyLength)
        
        let passwordPtr = [UInt8](passwordData)
        let status = CCKeyDerivationPBKDF(
            CCPBKDFAlgorithm(kCCPBKDF2),
            passwordPtr, passwordData.count,
            salt, salt.count,
            CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
            iterations,
            &derivedKey, keyLength
        )
        
        guard status == kCCSuccess else {
            throw NSError(domain: "KeyManager", code: Int(status), userInfo: [NSLocalizedDescriptionKey: "Key derivation failed"])
        }
        
        return derivedKey
    }
    
    private func encryptWithKEK(payload: [UInt8], kek: [UInt8]) throws -> String {
        let symmetricKey = SymmetricKey(data: kek)
        let ivData = [UInt8]((0..<12).map { _ in UInt8.random(in: 0...255) })
        let nonce = try AES.GCM.Nonce(data: ivData)
        
        let sealedBox = try AES.GCM.seal(payload, using: symmetricKey, nonce: nonce)
        let ciphertext = sealedBox.ciphertext
        // Tag is part of the sealedBox but we need to append it. Android does cipher.doFinal() which appends the 16-byte tag.
        // In CryptoKit, sealedBox.combined returns Nonce + Ciphertext + Tag. Wait, no.
        // Android AES/GCM/NoPadding appends the tag to the ciphertext.
        // So Android combined = IV (12) + Ciphertext + Tag (16)
        
        var combined = [UInt8]()
        combined.append(contentsOf: ivData)
        combined.append(contentsOf: ciphertext)
        combined.append(contentsOf: sealedBox.tag)
        
        return Data(combined).base64EncodedString()
    }
    
    private func decryptWithKEK(encryptedBase64: String, kek: [UInt8]) throws -> [UInt8] {
        guard let combinedData = Data(base64Encoded: encryptedBase64) else {
            throw NSError(domain: "KeyManager", code: 1, userInfo: [NSLocalizedDescriptionKey: "Invalid base64 payload"])
        }
        
        let combined = [UInt8](combinedData)
        if combined.count < 12 + 16 { // IV + Tag
            throw NSError(domain: "KeyManager", code: 2, userInfo: [NSLocalizedDescriptionKey: "Invalid encrypted payload size"])
        }
        
        let iv = [UInt8](combined[0..<12])
        let ciphertextAndTag = [UInt8](combined[12...])
        
        // Split ciphertext and tag
        let ciphertext = [UInt8](ciphertextAndTag[0..<(ciphertextAndTag.count - 16)])
        let tag = [UInt8](ciphertextAndTag[(ciphertextAndTag.count - 16)...])
        
        let symmetricKey = SymmetricKey(data: kek)
        let nonce = try AES.GCM.Nonce(data: iv)
        let sealedBox = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ciphertext, tag: tag)
        
        let decryptedData = try AES.GCM.open(sealedBox, using: symmetricKey)
        return [UInt8](decryptedData)
    }
    
    func hasV2Key() -> Bool {
        return keychainGet(key: ENCRYPTED_MASTER_KEY_V2) != nil
    }
    
    func generateRandomMasterKeyAndSave(pin: String) throws -> [UInt8] {
        var masterKey = [UInt8](repeating: 0, count: 32)
        _ = SecRandomCopyBytes(kSecRandomDefault, 32, &masterKey)
        
        let kek = try deriveLocalKEK(pin: pin)
        let encrypted = try encryptWithKEK(payload: masterKey, kek: kek)
        try keychainSet(key: ENCRYPTED_MASTER_KEY_V2, value: encrypted)
        
        return masterKey
    }
    
    func getMasterKeyForPin(pin: String) throws -> [UInt8] {
        if let v2Encoded = keychainGet(key: ENCRYPTED_MASTER_KEY_V2) {
            let kek = try deriveLocalKEK(pin: pin)
            return try decryptWithKEK(encryptedBase64: v2Encoded, kek: kek)
        }
        // Fallback V1
        return try deriveLocalKEK(pin: pin)
    }
    
    func upgradeToV2(masterKey: [UInt8], pin: String) throws {
        if keychainGet(key: ENCRYPTED_MASTER_KEY_V2) == nil {
            let kek = try deriveLocalKEK(pin: pin)
            let encrypted = try encryptWithKEK(payload: masterKey, kek: kek)
            try keychainSet(key: ENCRYPTED_MASTER_KEY_V2, value: encrypted)
        }
    }
    
    func exportCloudMasterKey(pin: String) throws -> String {
        let masterKey = try getMasterKeyForPin(pin: pin)
        let exportKEK = try deriveExportKEK(pin: pin)
        return try encryptWithKEK(payload: masterKey, kek: exportKEK)
    }
    
    func importCloudMasterKey(payload: String, pin: String) throws {
        let exportKEK = try deriveExportKEK(pin: pin)
        let masterKey = try decryptWithKEK(encryptedBase64: payload, kek: exportKEK)
        
        let localKEK = try deriveLocalKEK(pin: pin)
        let encryptedLocal = try encryptWithKEK(payload: masterKey, kek: localKEK)
        try keychainSet(key: ENCRYPTED_MASTER_KEY_V2, value: encryptedLocal)
        
        try storeMasterKey(key: masterKey)
    }
    
    func verifyCloudMasterKeyMatch(payload: String, pin: String) -> Bool {
        do {
            let exportKEK = try deriveExportKEK(pin: pin)
            let cloudMasterKey = try decryptWithKEK(encryptedBase64: payload, kek: exportKEK)
            let currentMasterKey = try getMasterKeyForPin(pin: pin)
            return cloudMasterKey == currentMasterKey
        } catch {
            return false
        }
    }
    
    func changePinV2(oldPin: String, newPin: String) throws {
        let masterKey = try getMasterKeyForPin(pin: oldPin)
        let newLocalKEK = try deriveLocalKEK(pin: newPin)
        let encryptedLocal = try encryptWithKEK(payload: masterKey, kek: newLocalKEK)
        try keychainSet(key: ENCRYPTED_MASTER_KEY_V2, value: encryptedLocal)
    }
    
    func storeMasterKey(key: [UInt8]) throws {
        let encodedKey = Data(key).base64EncodedString()
        try keychainSet(key: KEY_ALIAS, value: encodedKey)
    }
    
    func getMasterKey() -> [UInt8]? {
        guard let encodedKey = keychainGet(key: KEY_ALIAS),
              let keyData = Data(base64Encoded: encodedKey) else {
            return nil
        }
        return [UInt8](keyData)
    }
    
    func encrypt(plaintext: String, key: [UInt8]? = nil) throws -> String {
        let activeKey = key ?? getMasterKey()
        guard let activeKey = activeKey else { throw NSError(domain: "KeyManager", code: 3, userInfo: [NSLocalizedDescriptionKey: "No master key available"]) }
        
        guard let plaintextData = plaintext.data(using: .utf8) else { throw NSError(domain: "KeyManager", code: 4, userInfo: [NSLocalizedDescriptionKey: "Invalid string"]) }
        return try encryptWithKEK(payload: [UInt8](plaintextData), kek: activeKey)
    }
    
    func decrypt(encryptedData: String, key: [UInt8]? = nil) throws -> String {
        let activeKey = key ?? getMasterKey()
        guard let activeKey = activeKey else { throw NSError(domain: "KeyManager", code: 3, userInfo: [NSLocalizedDescriptionKey: "No master key available"]) }
        
        let plaintextBytes = try decryptWithKEK(encryptedBase64: encryptedData, kek: activeKey)
        guard let plaintext = String(data: Data(plaintextBytes), encoding: .utf8) else { throw NSError(domain: "KeyManager", code: 5, userInfo: [NSLocalizedDescriptionKey: "Could not decode string"]) }
        return plaintext
    }
    
    func clear() {
        keychainDelete(key: KEY_ALIAS)
    }
    
    func clearAll() {
        clear()
        keychainDelete(key: ENCRYPTED_MASTER_KEY_V2)
        // Salt lives in UserDefaults and is wiped on uninstall, but must be cleared
        // explicitly when resetting encryption state (matches Android key wipe behavior).
        defaults.removeObject(forKey: SALT_KEY)
    }
    
    // MARK: - Keychain Helpers
    
    private func keychainSet(key: String, value: String) throws {
        guard let data = value.data(using: .utf8) else { return }
        
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: key,
            kSecValueData as String: data,
            kSecAttrAccessible as String: kSecAttrAccessibleWhenUnlockedThisDeviceOnly
        ]
        
        SecItemDelete(query as CFDictionary)
        let status = SecItemAdd(query as CFDictionary, nil)
        
        if status != errSecSuccess {
            throw NSError(domain: "KeyManager", code: Int(status), userInfo: [NSLocalizedDescriptionKey: "Keychain write failed"])
        }
    }
    
    private func keychainGet(key: String) -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: key,
            kSecReturnData as String: kCFBooleanTrue!,
            kSecMatchLimit as String: kSecMatchLimitOne
        ]
        
        var dataTypeRef: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &dataTypeRef)
        
        if status == errSecSuccess, let data = dataTypeRef as? Data {
            return String(data: data, encoding: .utf8)
        }
        return nil
    }
    
    private func keychainDelete(key: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: key
        ]
        SecItemDelete(query as CFDictionary)
    }
}
